Add API for Headers on Parts of Multipart Form Data

jcaesar opened the pull request:
When working on [rust-ipfs-api](https://github.com/gkbrk/rust-ipfs-api), I saw that it uses a somewhat uncommon format for some requests:
```
POST /api/v0/add?chunker=size-262144&encoding=json&hash=sha2-256&nocopy=true&pin=false&progress=true&recursive=false&stream-channels=true HTTP/1.1
Host: 127.0.0.1:5001
User-Agent: go-ipfs-cmds/http
Connection: close
Transfer-Encoding: chunked
Content-Type: multipart/form-data; boundary=cfc4c1c4ad2e8ad270952524bc619a97033c8e6429a52d1358992325ed8b
Accept-Encoding: gzip

--cfc4c1c4ad2e8ad270952524bc619a97033c8e6429a52d1358992325ed8b
Abspath: /home/ipfs/gentoo/releases/verify-digests.sh
Content-Disposition: file; filename="verify-digests.sh"
Content-Type: application/octet-stream

#!/bin/bash
# Name: verify-digests.sh
…
```
The `Abspath` header on the form data part is critical: it tells the server where to find the file on hard disk if it needs to re-read it later. (Whether that is good design…)

This PR does add a patch that allows composing requests like the above, but considering my unfamiliarity with Rust and the discussion in https://github.com/seanmonstar/reqwest/issues/312 I doubt that you want to accept it like this. If possible, I'd like you to consider this a feature request, and a request to tell me how to implement the functionality properly.

seanmonstar commented:
Being able to set headers for a part seems reasonable. I'd suggest that instead of a `HashMap`, instead just have it be an `http::HeaderMap`.

jcaesar commented:
I've switched to `http::HeaderMap` and added a little test. I noticed that this may cause a re-appearance of https://github.com/hyperium/hyper/issues/1492 Not sure what to do about that.

seanmonstar commented on src/multipart_.rs:
```diff
@@ -7,6 +7,10 @@ use std::path::Path;
+use http;
+
+/// An alias to http::HeaderMap so http is unneded as an explicit dependency
+pub type HeaderMap = http::HeaderMap<String>;
```
I'd remove this alias, and use the default `http::HeaderMap` which uses `HeaderValue` instead of `String`.

seanmonstar commented on src/multipart_.rs:
```diff
@@ -197,6 +203,21 @@ impl Part {
     }
+
+    /// Returns a reference to the map of additional header fields
+    pub fn header_fields(&self) -> &HeaderMap {
```
To match with the other APIs, I'd call this method just `headers`.

seanmonstar commented on src/multipart_.rs:
```diff
@@ -197,6 +203,21 @@ impl Part {
+    }
+
+    /// Returns a mutable reference to the map of additional header fields
+    pub fn header_fields_mut(&mut self) -> &mut HeaderMap {
```
And this method could be named `headers_mut`.

seanmonstar commented on src/multipart_.rs:
```diff
@@ -197,6 +203,21 @@ impl Part {
+    }
+
+    /// Replaces the additional header fields
+    pub fn replace_header_fields(&mut self, hdr: HeaderMap) -> () {
```
This method shouldn't be needed, as anyone can do the same with `std::mem::replace` combined with `headers_mut`.

seanmonstar commented on src/multipart_.rs:
```diff
@@ -205,6 +226,7 @@ impl fmt::Debug for Part {
             .field("value", &self.value)
             .field("mime", &self.mime)
             .field("file_name", &self.file_name)
+            .field("hdr", &self.hdr)
```
For debugging, we may as well call it the full word, `headers`.

jcaesar replied:
I renamed the struct field, too…

seanmonstar commented on src/multipart_.rs:
```diff
@@ -314,7 +336,12 @@ fn header(name: &str, field: &Part) -> String {
+        field.hdr.iter().fold(
+            "".to_string(),
+            |hdrs, (k,v)|
+                format!("{}\r\n{}: {}", hdrs, k.as_str(), v)
```
This will actually allocate a new string for each header, whereas they could just be pushed onto the existing one... Though, there shouldn't be that many for a single part, so it's probably not a big deal.

jcaesar replied:
I hope + on String does what I think it does.

jcaesar commented:
Review implemented and rebased on master (with `pub(crate)`). I hope I got the rebase right…

jcaesar commented:
Thank you for merging and enhancing it. (And sorry for the delay. I was moving over the weekend.)
